Source/Core/Core/HW/EXI/BBA/XLINK_KAI_BBA.cpp

```diff
@@ -0,0 +1,277 @@
+          // u32(bytes_read - 4), 0x10).c_str());
+          self->m_eth_ref->mRecvBufferLength = u32(bytes_read - 4);
+          self->m_eth_ref->RecvHandlePacket();
+        }
+      }
+    }
+    // Otherwise we recieved control data or junk
+    else
+    {
+      std::string control_msg(self->m_in_frame, self->m_in_frame + bytes_read);
+      INFO_LOG(SP1, "Received XLink Kai control data: %s", control_msg.c_str());
+
+      // connected;identifier
+      if (StringBeginsWith(control_msg, "connected"))
+      {
+        NOTICE_LOG(SP1, "XLink Kai BBA connected");
+        OSD::AddMessage("XLink Kai BBA connected", 4500);
+
+        // Send any extra settings now
+        // Enable XLink chat messages in OSD if set
+        if (self->m_chat_osd_enabled)
+        {
+          constexpr std::string_view cmd = "setting;chat;true";
+          const auto size = u32(cmd.length());
+          u8 buffer[255] = {};
+          memmove(buffer, cmd.data(), size);
+
+          DEBUG_LOG(SP1, "SendCommandPayload %x\n%s", size,
+                    ArrayToString(buffer, size, 0x10).c_str());
+
+          if (self->m_sf_socket.send(buffer, size, self->m_sf_recipient_ip, self->m_dest_port) !=
+              sf::Socket::Done)
+          {
+            ERROR_LOG(SP1,
+                      "ReadThreadHandler(): failed to send setting message to XLink Kai client");
+          }
+        }
+      }
+      // disconnected;optional_identifier;optional_message
+      else if (StringBeginsWith(control_msg, "disconnected"))
+      {
+        NOTICE_LOG(SP1, "XLink Kai BBA disconnected");
+        // Show OSD message for 15 seconds to make sure the user sees it
+        OSD::AddMessage("XLink Kai BBA disconnected", 15000);
+
+        // Disable socket blocking
+        self->m_sf_socket.setBlocking(false);
+
+        // Shut down the read thread
+        self->readEnabled.Clear();
+        self->readThreadShutdown.Set();
+
+        // Close the socket
+        self->m_sf_socket.unbind();
+        break;
+      }
+      // keepalive;
+      else if (StringBeginsWith(control_msg, "keepalive"))
+      {
+        DEBUG_LOG(SP1, "XLink Kai BBA keepalive");
+
+        // Only uncomment for debugging, just clogs the log otherwise
+        // INFO_LOG(SP1, "SendCommandPayload %x\n%s", 2, ArrayToString(m_in_frame, 2,
+        // 0x10).c_str());
+
+        // Reply (using the message that came in!)
+        if (self->m_sf_socket.send(self->m_in_frame, 10, self->m_sf_recipient_ip,
+                                   self->m_dest_port) != sf::Socket::Done)
+        {
+          ERROR_LOG(SP1, "ReadThreadHandler(): failed to reply to XLink Kai client keepalive");
+        }
+      }
+      // message;message_text
+      else if (StringBeginsWith(control_msg, "message"))
+      {
+        NOTICE_LOG(SP1, "XLink Kai message: %s",
+                   (control_msg.substr(8, control_msg.length() - 1)).c_str());
```

Comment:
Missed this in the last review (my bad), but you may want to place this in a local like so:

```cpp
std::string msg = control_msg.substr(8, control_msg.length() - 1);
```

as this is duplicated in the line below, which can cause some unnecessary string churn.